File: output_20260321_064953/qwen3/c/Struct/spouse.c
#include "stdlib.h"
#include "spouse.h"

/*@ predicate person(struct person* p; struct person* spouse) =
      p != 0 &*&
      malloc_block_person(p) &*&
      p->spouse |-> spouse;
@*/

struct person {
  char* name;
  struct person* spouse;
};

struct person *create_person()
//@ requires true;
//@ ensures person(result, 0);
{
  struct person *p = malloc(sizeof(struct person));
  if(p == 0) abort();
  p->spouse = 0;
  //@ close person(p, 0);
  return p;
}

void marry(struct person *this, struct person *other)
//@ requires person(this, _) &*& person(other, _);
//@ ensures person(this, other) &*& person(other, this);
{
  //@ open person(this, _);
  //@ open person(other, _);
  this->spouse = other;
  other->spouse = this;
  //@ close person(this, other);
  //@ close person(other, this);
}

struct person* person_get_spouse(struct person* this)
//@ requires person(this, ?s);
//@ ensures person(this, s) &*& result == s;
{
  //@ open person(this, s);
  struct person* sp = this->spouse;
  //@ close person(this, s);
  return sp;
}

void divorce(struct person* this)
//@ requires person(this, ?s) &*& s != 0 &*& person(s, this);
//@ ensures person(this, 0) &*& person(s, 0);
{
  //@ open person(this, s);
  //@ open person(s, this);
  this->spouse->spouse = 0;
  this->spouse = 0;
  //@ close person(this, 0);
  //@ close person(s, 0);
}

void die(struct person *this)
//@ requires person(this, ?s) &*& (s == 0 ? true : person(s, this));
//@ ensures s == 0 ? true : person(s, 0);
{
  if(this->spouse != 0) {
    //@ open person(this, s);
    //@ open person(s, this);
    this->spouse->spouse = 0;
    //@ close person(s, 0);
  } else {
    //@ open person(this, 0);
  }
  free(this); 
}

int main() 
//@ requires true;
//@ ensures true;
{
  struct person* alice = create_person();
  struct person* bob = create_person();
  struct person* eve = 0;
  marry(alice, bob);
  eve = create_person();
  divorce(bob);
  marry(bob, eve);
  die(alice);
  die(bob);
  die(eve);
  return 0;
}

File: output_at_5/qwen3/sample_1/java/java/Linklist/Spouse.java
class Person {

    protected Person spouse;
    
    /*@
    predicate person(Person p) = true;
    @*/

    public  void spouse_symm()
        
        
    {
        //@ requires person(this);
        //@ ensures person(this);
        
    }

    public Person()
        
        
    {
        //@ requires true;
        //@ ensures person(this) &*& this.spouse == null;
        
    }
    
    public Person getSpouse()
        
        
    {
        //@ requires person(this);
        //@ ensures person(this) &*& result == this.spouse;
        
        return spouse;
        
    }
    
    protected void setSpouse(Person other)
        
        
    {
        //@ requires person(this) &*& person(other);
        //@ open person(this);
        //@ open person(other);
        //@ ensures person(this) &*& person(other);
        //@ close person(this);
        //@ close person(other);
        
        spouse = other;
        other.spouse = this;
        
    }
    
    protected void clearSpouse()
        
        
    {
        //@ requires person(this) &*& spouse != null;
        //@ open person(this);
        //@ open person(spouse);
        //@ ensures person(this) &*& person(spouse) &*& this.spouse == null;
        //@ close person(this);
        //@ close person(spouse);
        
        spouse.spouse = null;
        spouse = null;
        
    }
    
    void marry(Person other)
        
        
    {
        //@ requires person(this) &*& person(other);
        //@ open person(this);
        //@ open person(other);
        //@ ensures person(this) &*& person(other);
        //@ close person(this);
        //@ close person(other);
        
        other.setSpouse(this);
        
    }
    
    void divorce()
        
        
    {
        //@ requires person(this) &*& spouse != null;
        //@ open person(this);
        //@ open person(spouse);
        //@ ensures person(this) &*& person(spouse) &*& this.spouse == null;
        //@ close person(this);
        //@ close person(spouse);
        
        spouse.clearSpouse();
        
    }

}

class Program {

    public static void main(String[] args)
        
        
    {
        Person a = new Person();
        Person b = new Person();
        a.marry(b);
        b.divorce();
    }

}